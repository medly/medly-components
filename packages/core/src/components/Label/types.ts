import { FontVariants, FontWeights } from '@medly-components/theme';
import { HTMLProps } from '@medly-components/utils';

export type LabelPositions = 'top' | 'right' | 'left' | 'bottom';

export interface LabelCommonProps extends HTMLProps<HTMLLabelElement> {
    /** Shows pointer while hovering on it
     * @default false
     */
    showPointer?: boolean;
    /** Label position
     * @default left
     */
    labelPosition?: LabelPositions;
    /** Shows red color asterisk
     * @default false
     */
    required?: boolean;
    /** Label Color */
    labelColor?: string;
}

export interface LabelProps extends LabelCommonProps {
    /** Label Variant */
    labelVariant?: FontVariants;
    /** Label Weight */
    labelWeight?: FontWeights;
}

export type StyledLabelProps = LabelProps & {
    labelPosition: LabelPositions;
};
