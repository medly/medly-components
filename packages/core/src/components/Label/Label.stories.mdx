import { Label } from './Label';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Label.stories';

<Meta title="Core" component={Label} />

# Label

The `Label` component lets you add captions to any element in your project. You can click or tap the label to activate the associated element.

<Preview withToolbar>
    <Story name="Label">{stories.Basic()}</Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Label} />

## LabelTheme

This component offers the following themes mentioned in the table below:

<Props of={stories.ThemeInterface} />
