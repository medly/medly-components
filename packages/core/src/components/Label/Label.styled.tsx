import { defaultTheme } from '@medly-components/theme';
import { css, oppositePositionalSpacing, styled, WithThemeProp } from '@medly-components/utils';
import { StyledLabelProps } from './types';

const asterisk = () => css`
    ::after {
        color: red;
        content: ' *';
    }
`;

const getTextStyle = ({ theme, ...props }: StyledLabelProps & WithThemeProp) => {
    const { defaults } = theme.label,
        { weights, variants } = theme.font,
        { labelVariant, labelColor, labelWeight } = props,
        { fontSize, fontWeight, lineHeight, letterSpacing } = variants[labelVariant || defaults.variant];

    return css`
        color: ${labelColor || defaults.color};
        font-size: ${fontSize};
        font-weight: ${weights[labelWeight || fontWeight]};
        letter-spacing: ${letterSpacing};
        line-height: ${lineHeight};
    `;
};

export const LabelStyled = styled('label')<StyledLabelProps>`
    ${getTextStyle}
    ${({ required }) => required && asterisk()}
    margin: ${({ theme, labelPosition }) => oppositePositionalSpacing(labelPosition, theme.spacing.S2)};
    cursor: ${({ showPointer }) => (showPointer === undefined ? 'default' : showPointer ? 'pointer' : 'not-allowed')};
`;

LabelStyled.defaultProps = {
    theme: defaultTheme
};
