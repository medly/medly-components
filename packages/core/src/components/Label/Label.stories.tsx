import { defaultTheme, FontVariants, LabelTheme } from '@medly-components/theme';
import { boolean, color, select } from '@storybook/addon-knobs';
import { Label } from './Label';
import type { FC } from 'react';

const variantOptions: FontVariants[] = ['h1', 'h2', 'h3', 'h4', 'h5', 'body1', 'body2'];
const defaultColor = defaultTheme.font.defaults.color;

export const ThemeInterface: FC<LabelTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.label
};

export const Basic = () => (
    <Label
        required={boolean('Show Required asterisk', false)}
        labelColor={color('Color', defaultColor)}
        labelVariant={select('Label Variant', variantOptions, 'body2')}
    >
        Top 5
    </Label>
);
