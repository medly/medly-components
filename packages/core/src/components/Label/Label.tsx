import { WithStyle } from '@medly-components/utils';
import React, { FC } from 'react';
import { LabelStyled } from './Label.styled';
import { LabelProps } from './types';

const Component: FC<LabelProps> = React.memo(
    React.forwardRef((props, ref) => (
        <LabelStyled ref={ref} labelPosition={props.labelPosition!} {...props}>
            {props.children}
        </LabelStyled>
    ))
);

Component.defaultProps = {
    labelPosition: 'left',
    required: false
};
Component.displayName = 'Label';
export const Label: FC<LabelProps> & WithStyle = Object.assign(Component, { Style: LabelStyled });
