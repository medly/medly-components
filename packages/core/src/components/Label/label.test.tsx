import { render } from '@test-utils';
import { Label } from './Label';

describe('Label component', () => {
    it('should render correctly with default props', () => {
        const { container } = render(<Label>Label</Label>);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly with all the props given', () => {
        const { container } = render(
            <Label required showPointer labelPosition="top" labelVariant="body2" labelColor="pink">
                Demo Label
            </Label>
        );
        expect(container).toMatchSnapshot();
    });
});
