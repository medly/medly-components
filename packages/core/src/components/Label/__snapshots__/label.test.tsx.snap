// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label component should render correctly with all the props given 1`] = `
.c0 {
  color: pink;
  font-size: 1.4rem;
  font-weight: 400;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0px 0.8rem 0px;
  cursor: pointer;
}

.c0 ::after {
  color: red;
  content: ' *';
}

<div>
  <label
    class="c0"
    labelcolor="pink"
    labelposition="top"
    labelvariant="body2"
    required=""
  >
    Demo Label
  </label>
</div>
`;

exports[`Label component should render correctly with default props 1`] = `
.c0 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0.8rem 0px 0px;
  cursor: default;
}

<div>
  <label
    class="c0"
    labelposition="left"
  >
    Label
  </label>
</div>
`;
